### DISCUSSION TITLE
ESP32 receiver soldering with F3 EVO

### DISCUSSION THREAD

**Discussion by clennpillo** [jyjblrd/Mocap-Drones#46] [General]
hi,

I got a problem that I dont' know how to soldering ESP32 receiver to F3 EVO board, I watched the video but there was no details about this.

My betaflight is as follow
![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/d93ca483-f154-43e5-a7aa-cdb01a23b13a)

### ESP32 receiver soldering
I suppose four terminals need to be soldered.
![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/ed77cf42-8e9a-4513-9aea-fcf8d5d7a0e1)

### F3 EVO soldering
I know how to solder power terminals, but for signals, because my betaflight showing that UART2's serial function is enabled as in the screen shot above, I guess I need to use UART2's two terminals, one is TX, another is RX, however, I couldn't find any article or video that explain which terminals are for UART1, UART2, UART3.
![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/1fbe11a0-022c-4e7a-9be4-c48b494b8a31)

I tried to connect in the way above before soldering, but it turned out the motors won't start, so I think there might be some problem here.

Can you tell me where can I find this information about UART1,2,3 's corresponding terminals, and can you give me your sodering picture ?

**Comment by cultist123** [comment]
Although I haven't solved this problem for you, I have done some research on it. The F3 board does not have the corresponding communication pins, so you can only solder it directly to the STM32 chip. I suggest you refer to the STM32 chip diagram as shown below, and solder it as demonstrated in the video.

<img width="1546" alt="截屏2024-05-26 20 43 44" src="https://github.com/jyjblrd/Low-Cost-Mocap/assets/121796670/c8ce1e83-c515-4816-9c23-9c2a7b99a718">

I guess data transmission is only one-way on STM32, it only sends data and doesn't receive.
[
<img width="1500" alt="截屏2024-05-26 20 48 53" src="https://github.com/jyjblrd/Low-Cost-Mocap/assets/121796670/f95cf4ca-4d01-4dd8-868f-4ce615246533">
](url)
<img width="1500" alt="截屏2024-05-26 20 48 58" src="https://github.com/jyjblrd/Low-Cost-Mocap/assets/121796670/62ff4b08-f0b9-4f80-ac50-b4593d5bee65">

As shown in the diagram below, I believe there are only three connections in total, not four. However, I haven't verified it yet.
<img width="1500" alt="截屏2024-05-26 20 49 27" src="https://github.com/jyjblrd/Low-Cost-Mocap/assets/121796670/c42234f2-79f7-4e6a-a3e3-90169a7b9f54">

**Reply by jyjblrd** [reply]
Just to clarify on this comment, we do not need to connect the flight controller UART 3 TX pin to the ESP 32 (ie. FC --> ESP 32 communication is not needed).

**Reply by cultist123** [reply]
I noticed that you connected a resistor and other pins on the ESP32. According to my idea, only three wires are needed between F3 and ESP32. I want to know if I missed something.

**Comment by jyjblrd** [comment]
You actually need to connect the ESP32 to the UART 3. UART 1 is used by the usb interface (i think?) and UART 2 is used by the onboard RC receiver (trying to connect another signal to this pin didn't work for me, there prob is interference from the other RC receiver trying to also send signals on that UART)

[This location in the video](https://www.youtube.com/watch?v=0ql20JKrscQ&t=157s) shows the UART 3 RX pin to solder to. You can then tell betaflight to use UART 3 as the RC signal. It isn't the best solution, a better flight controller that actually breaks out the extra UART as a solderable pin would be better.

**Comment by clennpillo** [comment]
Thanks for jyjblrd and cultist123's replies which helps me out.

now I get it, the right terminal is on STM32 instead of F3 board. I just tried to connect as follows.

![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/5234071c-b438-49c3-b899-7a75f3b0c7c9)

on the ESP32 receiver side, I deployed React frontend, python backend ESP32 sender and ESP32 receiver, the receiver was able to send data via sbus and I could see the logs while I click buttons at React frontend, I changed some code slightly to let the code run even if the cameras are not ready yet. The screenshot is what I see in Arduino running ESP32 receiver.
![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/343e5c0d-d34d-40ff-9eb2-3b3d7945b2fa)


I clicked arm at frontend, but motors still won't start. Any suggestions ?

**Comment by clennpillo** [comment]
I did some debug, when I type status command in CLI of betaflight, I got RXLOSS which means the receiver is not ready, didn't operation connecting ESP32 to STM32 make a good receiver for the F3 EVO board ? 

![image](https://github.com/jyjblrd/Low-Cost-Mocap/assets/19600126/eddb47f2-609d-4a54-ad50-48193161d86f)

**Comment by jyjblrd** [comment]
Yes, RX loss means that the flight controller is not receiving any RC signal from the ESP32.
1. Are you sure that the cable is connected correctly to the flight controller's pin? It is a very very small pin as I'm sure you know.
2. Is the ESP 32 receiver's code configured to send the data out of the correct digital out pin?
3. Is Betaflight correctly configured to receive the SBus data from the ESP 32? The correct UART needs to be configured and the correct RC protocol type also needs to be configured.

In the "receiver" page in betaflight, you will be able to see if the RC signal is being correctly received by the flight controller.

**Comment by clennpillo** [comment]
Turned out I didn't make the tiny pin from STM32 connection right.

I did it again and now the RXLOSS problem is solved, so I guess I know the correct pair that should be soldered between F3 EVO and ESP32 receiver, however, I tried many many times to solder the tiny pin and failed many many times, that's really the most difficult part in this project, till now I 'm still trying.

**Comment by jyjblrd** [comment]
Yeah, a better flight controller which actually breaks out the extra UART would make things a lot easier. Better yet, the esp32 could be the flight controller itself (one core receiving and decoding messages, another core controlling the motors).

To be honest I did not consider that people would try reimplement this project, I posted it more as a reference to others who wanted to build something similar. As you are probably discovering, it is not exactly a plug-and-play project, a fair bit of troubleshooting will be involved.

**Comment by clennpillo** [comment]
This is a great project that exactly meet our needs, so I tried my best to use what hardware and software you use in order to understand this project, after making this right, we gonna make changes and upgrade,etc.  Really appreciate what you have done.

**Reply by cultist123** [reply]
Please provide a contact method for discussing hardware usage. I am Chinese, you can give me your WeChat.

**Reply by jyjblrd** [reply]
It would be good if discussions were kept on github so others can benefit from them. Or perhaps just make a post in the [show and tell](https://github.com/jyjblrd/Low-Cost-Mocap/discussions/categories/show-and-tell) discussion board if you get it working so that others can learn from your setup :)

**Reply by cultist123** [reply]
Yes, you're right,l didn't think about it before.Perhaps this is the spirit of open source code.Thank for your reminder.

**Reply by clennpillo** [reply]
finally I got the tiny pin soldering working by asking for professional help, cost me 40 Chinese yuan, I can now focus on cameras. @cultist123
